### DISCUSSION TITLE
Using Directories as Hard Drives?

### DISCUSSION THREAD

**Discussion by KPreston2900050** [86Box/86Box#1722] [Ideas]
Is this even possible?
No documentation makes it hard to know about features.

I have thousands of old dos programs on older hard drives that I can just connect through a USB hub but it would be a pain to create image and for what when DOS already has that capability of reading directories.

Dosbox is apples and oranges but would that not be a sensible feature to have?

**Comment by bachimusprime** [comment]
This will only ever be done if someone can reverse engineer the way VPC
does it, and even then, it has to be an option, because that stuff's REALLY
inaccurate.

**Comment by KPreston2900050** [comment]
I guess it can be complicated despite DOS is a fundamental and necessary function of Windows.

So there is no dragging and dropping, no access to actual physical drives.
How do you personally transfer loose files without turning every single one them into an adf file or iso?

My first attempt was to use DOBOX and transfer them to a hard drive image then use the Image using in 86box but results are mixed.

It just seems an unnecessary and convoluted task that should not be when the whole purpose of the emulator in the first place is to run the very files people have difficulty getting access to.

I'm going to load in the 86box github environment into my vs2017 and see what I can do.

**Reply by charabaruk** [reply]
DOS was only "fundamental and necessary" for the non-NT Windows (i.e. 1.x-3.x, 9x, ME). With NT, 2000, XP and onwards, DOS environments have always been emulated instead, whether through Microsoft's own built-in NTVDM in 32-bit builds of NT family Windows versions, or through software like 86Box.

**Comment by bachimusprime** [comment]
@KPreston2900050 Be careful, 86box probably doesn't compile in vs2017 yet. Also, I do turn every single one of them into an img or iso file lol.

**Comment by KPreston2900050** [comment]
I feel your pain but now we get into an even bigger issue of being spoiled.
I now use Windows 10 and before that I trace all the way back to XP 64 with NTFS because of the file limitations of dos.

I go even further because I have access to several thousands of pieces of Dos software which if I want to use 86Box then I have to have several smaller hard drive images and transfer thousands if not hundreds of thousands of files.

Dosbox just mounts the directory.
I'm not nagging because I think 86Box, PCEM and the knock off are more versatile in configurations.

It's this one inability that is hindering it to be the best.  Like you can have a steak off the grill and it's wonderful but you can make it great just by adding a dash of pepper.

I just got VS2017 because it imports from Github with little issues, I never thought about the compiling part.  I have work to do.

**Comment by OBattler** [comment]
The thing is, DOSBox emulates hard disks at the level of INT 13h calls - ie. it simulates a BIOS on its own. It also simulates DOS on its own. That's why it can do simulation of high-level things such as logical drives and directories with ease. 86Box and PCem emulate real hardware and use real BIOS'es and real operating system, so DOSBox-like folder sharing would be impossible to do.
This leaves us with two possible solutions - reverse engineer the folder sharing system of an existing product (VirtualBox, VMWare, Virtual PC 2007) and implement it, or make our own. The former depends on a lot of factors (VirtualBox is open-source, but its folder sharing might not support older operating systems, VMWare and Virtual PC 2007 do support eg. DOS but are proprietary so disassembly and reverse engineering would be needed), and the latter requires the effort to write our own driver for each and every single platform people may want to run.
This is why, for how much I'd love to hare folder sharing in 86Box, I haven't done it, and is why Sarah Walker hasn't done it for PCem, either.

**Reply by Torinde** [reply]
For DOS - there is [VBSF](https://git.javispedro.com/cgit/vbados.git/about/) utilizing the VirtualBox interface.

**Comment by eton975** [comment]
Could it basically be summed up as being a high-level vs low-level thing ie. the abstract concept of folders and files vs the more-real, dirty, low-level concept of bytes on a disk arranged in sectors? Where because DOSBOX doesn't just implement the PC hardware but also the DOS OS that runs on top of it, it can much more easily do some nifty little tricks to allow you to drag-and-drop into its virtual HDDs without issue than other emulators?

And because DOSBOX emulates a full DOS system with OS in its own code, the devs of DOSBOX can do all sorts of little tricks to make that high-level abstraction work seamlessly, whereas PCem/86box just emulate the hardware that DOS/Windows/old Linux/any PC OS work with.

Thus to do the same in this emulator would require careful research and coding to properly implement reading/writing from a modern folder structure? without serious data corruption or disk emulation accuracy problems resulting from not having a perfect representation of a disk's sectors? ie. how am i gonna get files to read properly at the ide controller command level when the files aren't represented in a file table in a flat disk image? if i tell the disk (like some disk partitioners would) to write bytecode xxxxxx to sector yyyyyy, how am i gonna do that without an image?

**Comment by OBattler** [comment]
What you need to do is simulate a MBR with partition table, followed by a partition table with FAT's etc., and keep an array mapping paths to files to virtual sectors. Such a thing is best done as read-only. Writing per se is not difficult, but handling file deletion is.
I recommend looking at doing it as a SCSI device as I think SCSI controllers support disks that are just sequentially written files with no structure of any sort, but I need to double check that.

**Comment by eton975** [comment]
Nice way of putting it, cheers.

**Comment by MobyGamer** [comment]
An easier solution is to just write a network redirector TSR for the emulator.  It won't be able to boot x86 systems from a hosted filesystem, but you could certainly access the files from it.

**Comment by darkstar** [comment]
For now it's probably easiest to run LAN Manager and connect to your windows system over CIFS/SMB

**Comment by darkstar** [comment]
I just found this: [EtherDFS](http://etherdfs.sourceforge.net/). It might be a nice solution. 

It could be integrated (as optional BIOS extension for example, using a custom packet driver) with the server part running in the 86box host process and having both communicate via some trapdoor API call.

@MobyGamer might be interesting for you as well ;-)

**Comment by OBattler** [comment]
This will be looked at for 86Box v3.0. For time being, I'm closing this.

**Comment by cesss** [comment]
Let me add a vote for this feature. I'm looking for a DOS emulator with two requirements: A) Reasonably accurate CPU speed emulation (I don't really need scientifically-correct cycle accuracy, but I need that an emulated 486 at 50MHz runs at "more or less" the same speed as a real 486 at 50 MHz). The second requirement B) is to be able to mount a host folder as a hard drive in the emulator.

Most DOS emulators out there just run "at maximum speed" which turns them useless for recreating the experience of a real DOS PC at its real speed, and/or don't support mounting host folders as emulated drives. I think DOSBox is the only one currently supporting my requirements, but my success with it is mixed... some things work, other things don't. So, please consider implementing host folder mounts in 86Box, as it would possibly offer me everything I need in a DOS emulator.

**Comment by OBattler** [comment]
This is being looked into, though it's not going to be easy, as unlike DOSBox, 86Box only emulates the hardware and runs real BIOS ROM's and real OS'es on it, so it has no control over the DOS API.

**Comment by cesss** [comment]
I believe it would be possible to do a good implementation from QEMU's VVFAT (which Bochs uses as well). If I'm reading the docs correctly, the VVFAT hard disk looks like a low level image to the emulator, but however it's populated with the contents of a directory from the host when the emulator starts-up.

However, I'd say that VVFAT is "broken" in its current status, because the emulator is not able to update the virtual disk contents if you modify the files in the host from the host (the QEMU docs even mention that this can lead to undesired behaviour --not sure if they have improved this with the years). Moreover, if the emulator writes to the VVFAT disk, the writes are only pushed to the host when the emulator quits.

Thus, while the emulator is running, neither the host is allowed to write to the shared folder, nor the guest can send a file back to the host. This limits the usefulness of VVFAT a lot, and that's why I said it's "broken". Anyway, it's better than nothing.

However, I believe that the VVFAT implementation could be improved to workaround these limitations.

But... maybe it would be far easier just to patch DOS after installing it on 86Box, and redirect the DOS file I/O API to a virtualising interface within 86Box (yes, this approach might look against a clean pure hardware-only emulator implementation, but in fact that's exactly how VirtualBox does it: "patch" the guest OS -with installable extensions- that redirect stuff to an interface within VirtualBox). Maybe this approach would be easier than improving VVFAT. 

BTW, I also posted a related question at the VARCem issues list, asking if they have any means for shared folders, as I couldn't find anything at the docs.

**Comment by waltje** [comment]
Hi,

This is the whole problem with this feature.

The only 'real' option here is to (maybe based on VVFAT) implement a
removable hard disk, which is indeed populated by reading a folder's
contents upon mount (not startup.)

Windows, and many UNIX variants, have support for 'directory notifications',
which basically signal the application if a device or folder has changed. The
app will then un-mount the 'drive', re-populate it, and remount it.

From the guest, writing to the drive means the driver must immediately write
to the folder, which could be trickier to do.  The implementations I know do
this on a "close" file system operation.

Under DOS and Win3, this means installing an IFS module.

Regards,

Fred

--------------------------------------------
On Thu, 8/23/18, cesss <notifications@github.com> wrote:

 Subject: Re: [86Box/86Box] Using Directories as Hard Drives? (#187)
 To: "86Box/86Box" <86Box@noreply.github.com>
 Cc: "Subscribed" <subscribed@noreply.github.com>
 Date: Thursday, August 23, 2018, 5:32 AM
 
 I believe it would be possible to
 do a good implementation from QEMU's VVFAT (which Bochs
 uses as well). If I'm reading the docs correctly, the
 VVFAT hard disk looks like a low level image to the
 emulator, but however it's populated with the contents
 of a directory from the host when the emulator
 starts-up.
 However, I'd say that VVFAT is "broken" in
 its current status, because the emulator is not able to
 update the virtual disk contents if you modify the files in
 the host from the host (the QEMU docs even mention that this
 can lead to undesired behaviour --not sure if they have
 improved this with the years). Moreover, if the emulator
 writes to the VVFAT disk, the writes are only pushed to the
 host when the emulator quits.
 Thus, while the emulator is running, neither the host is
 allowed to write to the shared folder, nor the guest can
 send a file back to the host. This limits the usefulness of
 VVFAT a lot, and that's why I said it's
 "broken". Anyway, it's better than
 nothing.
 However, I believe that the VVFAT implementation could be
 improved to workaround these limitations.
 But... maybe it would be far easier just to patch DOS
 after installing it on 86Box, and redirect the DOS file I/O
 API to a virtualising interface within 86Box (yes, this
 approach might look against a clean pure hardware-only
 emulator implementation, but in fact that's exactly how
 VirtualBox does it: "patch" the guest OS -with
 installable extensions- that redirect stuff to an interface
 within VirtualBox). Maybe this approach would be easier than
 improving VVFAT.
 BTW, I also posted a related question at the VARCem
 issues list, asking if they have any means for shared
 folders, as I couldn't find anything at the docs.
 
 —
 You are receiving this because you are subscribed to this
 thread.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

**Comment by OBattler** [comment]
As for the suggestion to patch an OS with installable extensions, while I would personally love that option (that's what VMWare and Virtual PC do as well), it would require someone to write the required drivers for each and every single OS someone might want to do. That might be feasible for the large teams of VirtualBox, VMWare, or Virtual PC, which tend to be huge, but not for the small teams of 86Box or VARCem.

**Comment by rderooy** [comment]
A bit of an old issue, but just to mention that DOSBox-X with the latest release implemented (optionally) emulating host folders as harddisks for guest operating systems. So you can mount a host folder and when you boot the guest OS, such as MS-DOS or Win9x, it will dynamically emulate it as a FAT harddisk and it will appear like a regular HD in the guest OS.

Some limitations of the implementation for now:
- Cannot be used for the bootable drive
- One partition per drive
- Only HDD (no CD or Floppy)

https://dosbox-x.com/wiki/Guide%3AInstalling-Windows-98#_host_folder_mounts

**Comment by KPreston2900050** [comment]
7 years later, no solution ever implemented, world war 3 is a week away thanks to the Obama administration 3rd term Biden puppet and I am still getting this in my email.

The internet never forgets right...

**Reply by OBattler** [reply]
We did implement the ability to mount a folder as a CD-ROM. We can't do anything more than that - VVFAT's writability is currently a buggy mess.

**Reply by dhrdlicka** [reply]
We did not forget about anything, the previous posts precisely explain why this is a rather complex feature and why it's not as high on our bucket list.

There is an Unsubscribe button in the sidebar at the right if you don't wish to receive e-mail notifications about this.

**Comment by dhrdlicka** [comment]
I think it's worth mentioning that you can get pretty close to this by using native VHD mount in Windows 8 and above. If you use VHD disk images, you can double click them in File Explorer to mount them, making them appear as a real drive in Windows with a drive letter and everything. Just remember that the same image cannot be used by both Windows and 86Box at the same time, so make sure to unmount the image before launching 86Box.
